File: packages/backend-db/src/migrations/runner.ts
import { existsSync } from 'node:fs';
import { type DatabaseSync } from 'node:sqlite';
import {
  openDatabase,
  readDatabase,
  readSchemaVersion,
  rollbackQuietly,
  withDatabase,
  writeSchemaVersion,
} from '../common/index.js';
import { createBackup } from './backup.js';
import { createMigrationFailure } from './errors.js';
import { type Migration, type MigrationReport } from './types.js';

const probeVersion = (databasePath: string): number => {
  if (!existsSync(databasePath)) {
    return 0;
  }
  try {
    return readDatabase(databasePath, readSchemaVersion);
  } catch (cause) {
    throw createMigrationFailure({
      message: 'The database file could not be read and may be damaged.',
      cause,
    });
  }
};

const takeBackup = (
  database: DatabaseSync,
  databasePath: string,
  version: number,
): string => {
  try {
    return createBackup(database, databasePath, version);
  } catch (cause) {
    throw createMigrationFailure({
      message:
        'The database backup could not be created, so nothing was changed.',
      cause,
    });
  }
};

const applyMigration = (
  database: DatabaseSync,
  statements: Migration,
  version: number,
  backupPath: string | undefined,
): void => {
  try {
    database.exec('BEGIN IMMEDIATE');
    for (const statement of statements) {
      database.exec(statement);
    }
    const violations = database.prepare('PRAGMA foreign_key_check').all();
    if (violations.length > 0) {
      throw new Error(`foreign key violations: ${JSON.stringify(violations)}`);
    }
    writeSchemaVersion(database, version);
    database.exec('COMMIT');
  } catch (cause) {
    rollbackQuietly(database);
    throw createMigrationFailure({
      committedVersion: version - 1,
      backupPath,
      message: `Migration v${String(version)} did not finish. It was rolled back, so the database still holds schema v${String(version - 1)}.`,
      cause,
    });
  }
};

export const runMigrations = (
  databasePath: string,
  steps: readonly Migration[],
): MigrationReport => {
  const latest = steps.length;
  const fromVersion = probeVersion(databasePath);
  if (fromVersion > latest) {
    throw createMigrationFailure({
      message: `The database holds schema v${String(fromVersion)}, which this build does not know: it supports v${String(latest)}. Install the newer version again, or restore an older copy of the database.`,
    });
  }
  if (fromVersion === latest) {
    return { fromVersion, toVersion: latest };
  }
  return withDatabase(
    openDatabase(databasePath, { foreignKeys: false }),
    (database) => {
      const backupPath =
        fromVersion > 0
          ? takeBackup(database, databasePath, fromVersion)
          : undefined;
      for (let version = fromVersion + 1; version <= latest; version += 1) {
        applyMigration(database, steps[version - 1], version, backupPath);
      }
      return { fromVersion, toVersion: latest, backupPath };
    },
  );
};

File: packages/server/crates/db/src/runner.rs
use std::path::{Path, PathBuf};

use rusqlite::Connection;

use crate::{
    backup::create_backup,
    database::{
        OpenOptions, open_database, open_readonly, read_schema_version, write_schema_version,
    },
    failure::{BoxedError, MigrationFailure},
    schema::{MIGRATIONS, Migration},
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MigrationReport {
    pub from_version: usize,
    pub to_version: usize,
    pub backup_path: Option<PathBuf>,
}

fn probe_version(database_path: &Path) -> Result<usize, MigrationFailure> {
    if !database_path.exists() {
        return Ok(0);
    }
    open_readonly(database_path)
        .and_then(|connection| read_schema_version(&connection))
        .map_err(|cause| {
            MigrationFailure::new("The database file could not be read and may be damaged.")
                .caused_by(cause)
        })
}

fn take_backup(
    connection: &Connection,
    database_path: &Path,
    version: usize,
) -> Result<PathBuf, MigrationFailure> {
    create_backup(connection, database_path, version).map_err(|cause| {
        MigrationFailure::new("The database backup could not be created, so nothing was changed.")
            .caused_by(cause)
    })
}

fn run_statements(
    connection: &Connection,
    statements: Migration,
    version: usize,
) -> Result<(), BoxedError> {
    connection.execute_batch("BEGIN IMMEDIATE")?;
    for statement in statements {
        connection.execute_batch(statement)?;
    }
    let mut violations = connection.prepare("PRAGMA foreign_key_check")?;
    if violations.query([])?.next()?.is_some() {
        return Err("the migration left a dangling foreign key".into());
    }
    drop(violations);
    write_schema_version(connection, version)?;
    connection.execute_batch("COMMIT")?;
    Ok(())
}

fn apply_migration(
    connection: &Connection,
    statements: Migration,
    version: usize,
    backup_path: Option<&Path>,
) -> Result<(), MigrationFailure> {
    run_statements(connection, statements, version).map_err(|cause| {
        let _ = connection.execute_batch("ROLLBACK");
        let previous = version - 1;
        MigrationFailure::new(format!(
            "Migration v{version} did not finish. It was rolled back, so the database still holds schema v{previous}."
        ))
        .committed(previous)
        .backed_up(backup_path)
        .caused_by(cause)
    })
}

pub fn run_migrations(
    database_path: &Path,
    steps: &[Migration],
) -> Result<MigrationReport, MigrationFailure> {
    let latest = steps.len();
    let from_version = probe_version(database_path)?;
    if from_version > latest {
        return Err(MigrationFailure::new(format!(
            "The database holds schema v{from_version}, which this build does not know: it supports v{latest}. Install the newer version again, or restore an older copy of the database."
        )));
    }
    if from_version == latest {
        return Ok(MigrationReport {
            from_version,
            to_version: latest,
            backup_path: None,
        });
    }

    let connection = open_database(
        database_path,
        &OpenOptions {
            foreign_keys: false,
        },
    )
    .map_err(|cause| {
        MigrationFailure::new("The database could not be opened for migration.").caused_by(cause)
    })?;
    let backup_path = if from_version > 0 {
        Some(take_backup(&connection, database_path, from_version)?)
    } else {
        None
    };
    for version in (from_version + 1)..=latest {
        apply_migration(
            &connection,
            steps[version - 1],
            version,
            backup_path.as_deref(),
        )?;
    }
    Ok(MigrationReport {
        from_version,
        to_version: latest,
        backup_path,
    })
}

pub fn init_database(database_path: &Path) -> Result<MigrationReport, MigrationFailure> {
    run_migrations(database_path, MIGRATIONS)
}
